shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -0,0 +1,391 @@
+ */
+public class TextInputChannel {
+  @NonNull
+  public final MethodChannel channel;
+  @Nullable
+  private TextInputMethodHandler textInputMethodHandler;
+
+  private final MethodChannel.MethodCallHandler parsingMethodHandler = new MethodChannel.MethodCallHandler() {
+    @Override
+    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
+      if (textInputMethodHandler == null) {
+        // If no explicit TextInputMethodHandler has been registered then we don't
+        // need to forward this call to an API. Return.
+        return;
+      }
+
+      try {
+        String method = call.method;
+        Object args = call.arguments;
+        switch (method) {
+          case "TextInput.show":
+            textInputMethodHandler.show();
+            result.success(null);
+            break;
+          case "TextInput.hide":
+            textInputMethodHandler.hide();
+            result.success(null);
+            break;
+          case "TextInput.setClient":
+            final JSONArray argumentList = (JSONArray) args;
+            final int textInputClientId = argumentList.getInt(0);
+            final JSONObject jsonConfiguration = argumentList.getJSONObject(1);
+            textInputMethodHandler.setClient(textInputClientId, Configuration.fromJson(jsonConfiguration));
+            result.success(null);
+            break;
+          case "TextInput.setEditingState":
+            final JSONObject editingState = (JSONObject) args;
+            textInputMethodHandler.setEditingState(TextEditState.fromJson(editingState));
+            result.success(null);
+            break;
+          case "TextInput.clearClient":
+            textInputMethodHandler.clearClient();
+            result.success(null);
+            break;
+          default:
+            result.notImplemented();
+            break;
+        }
+      } catch (Exception exception) {
+        result.error("error", "JSON error: " + exception.getMessage(), null);
+      }
+    }
+  };
+
+  public TextInputChannel(@NonNull DartExecutor dartExecutor) {
+    this.channel = new MethodChannel(dartExecutor, "flutter/textinput", JSONMethodCodec.INSTANCE);
+    channel.setMethodCallHandler(parsingMethodHandler);
+  }
+
+  /**
+   * Instructs Flutter to update its text input editing state to reflect the given configuration.
+   */
+  public void updateEditingState(int inputClientId, String text, int selectionStart, int selectionEnd, int composingStart, int composingEnd) {
+    HashMap<Object, Object> state = new HashMap<>();
+    state.put("text", text);
+    state.put("selectionBase", selectionStart);
+    state.put("selectionExtent", selectionEnd);
+    state.put("composingBase", composingStart);
+    state.put("composingExtent", composingEnd);
+
+    channel.invokeMethod(
+        "TextInputClient.updateEditingState",
+        Arrays.asList(inputClientId, state)
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "newline" action.
```

Comment:
Are these actions supposed to be self explanatory? I'm not sure looking at the method signatures and docs what they actually do.

Reply:
They read ambiguously because in reality they are ambiguous.  The best we can describe here is the type of message or request being sent.  We have no idea or any guarantee that Flutter will take any particular action in response.  Moreover, the action that Flutter does take will likely be dependent on what user interaction is currently taking place, including what kind of text input the user is interacting with.  Does that make sense?